Use common polyfill for server and client, removing isomorphic-fetch

aduth opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #103054

## Proposed Changes

* Removes another instance of a `fetch` polyfill that can be removed due to [baseline support for `fetch` as of Node.js v21](https://nodejs.org/en/blog/announcements/v21-release-announce#stable-fetchwebstreams)
* Consolidates to a common entrypoint for polyfills, since the polyfill implementation is now identical between server and browser

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Eliminate unnecessary dependencies to reduce maintenance overhead and supply chain attack risk surface area

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

I wasn't able to find anywhere where we currently incur a `fetch` through server-side rendering, but I was able to verify with the introduction of a test route:

```diff
diff --git a/client/server/pages/index.js b/client/server/pages/index.js
index c69e132fa83..e1ed7f22e0d 100644
--- a/client/server/pages/index.js
+++ b/client/server/pages/index.js
@@ -999,2 +999,8 @@ export default function pages() {
 
+	app.get( '/test', async ( req, res ) => {
+		const result = await fetch( 'https://public-api.wordpress.com/geo/' );
+		const data = await result.json();
+		res.send( data );
+	} );
+
 	// Redirect legacy `/new` routes to the corresponding `/start`
```

After applied and starting the server, navigate to http://calypso.localhost:3000/test to verify you see the same output as from https://public-api.wordpress.com/geo/

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

aduth committed:
Use common polyfill for server and client, removing isomorphic-fetch

aduth commented on packages/calypso-polyfills/package.json:
```diff
@@ -2,8 +2,7 @@
 	"description": "Polyfills required for Calypso to work across various browsers and in node.",
-	"browser": "browser.js",
-	"main": "node.js",
+	"main": "index.js",
```
Aside: I was motivated to port this file to TypeScript, but it seemed like a lot of hassle to set up build scripts, etc. for the package, based on what I was seeing with other packages. I wonder if we'd be okay to point directly to `.ts` files without a build step, since Node.js now has [support for loading TypeScript files](https://nodejs.org/api/typescript.html#type-stripping)? 

(The history on that page is a little confusing, but I tested locally that it works without experimental flags or warnings in Node.js v22.9.0)

jsnajdr replied:
The polyfills package is ever used as a part of webpack build (in #103151 I just removed the only place where it mistakenly wasn't). It doesn't really make much sense anywhere else, because the `core-js/stable` import is expected to be further processed by Babel `preset-env` and expanded into a short list of individual polyfills that are actually needed by the target environment.

The webpack build should be able to consume the raw TS file just fine. Therefore you can just specify the simplest possible `exports` field:
```
"exports": "./index.ts"
```
and it will work.

(You'll also migrate away from legacy `main`/`module`/`browser` fields to the modern `exports` when doing this.)

aduth replied:
@jsnajdr Beyond how it's used internally, since this is a [published package](https://www.npmjs.com/package/@automattic/calypso-polyfills) I was cautious of the ramifications of porting to TypeScript, and also in a broad sense of using TypeScript in packages. With recent Node.js improvements for type-stripping TypeScript files, it seems like it should "just work" in relatively modern versions of Node.js and would make our lives a lot simpler.

jsnajdr replied:
I'm afraid it's published only formally, I'd be suprised if any other repo used it. It was published to NPM only once, by me in January 2020 🙂 

On the other hand, it's perfectly OK to keep the sources in JavaScript, as the entire package will now consist of a single `import` statement, there are no types around.

aduth replied:
Yeah I think I'll keep it as-is for now, though I'll be curious to experiment with other packages in the future to use TypeScript sources files without a build step. It feels like that would significantly simplify the developer experience and perhaps yield a performance improvement as well (at build time at least).

jsnajdr replied:
> experiment with other packages in the future to use TypeScript sources files without a build step.

Oh, but we already do this: that's what the `calypso:src` field is for. It points to the uncompiled TS/JS files, and tools like Jest or webpack can use the sources directly, because they already have a compilation pipeline set up to process these files.

What we could optimize is the `build-packages` step that generates the `dist` directories. That work is totally wasted most of the time, because the `dist` directories are useful only when publishing packages to NPM. And that's a very rare task. Everything else works with the `calypso:src` field and raw sources. I'm not sure about the TypeScript setup, maybe it still works with the `dist` files, but that's fixable: the TypeScript resolver can be configured with the `compilerOptions.customConditions` option.

aduth replied:
I'm more talking about publishing and consuming TypeScript files natively, without any sort of build step or transpilation (no `dist/` directories, no Jest or Webpack transpilation step). It should be possible to set `main` (`exports`) to an `index.ts` and it works without any other tools, even as a consumer of the published NPM package.

Internally I'm not entirely clear how useful it'd be, though it could be nice if we could get to a point where we don't have to use Jest or Webpack to preprocess at least some of our TypeScript files.

jsnajdr replied:
> I'm more talking about publishing and consuming TypeScript files natively

That would be breaking some formal/informal rules about what NPM packages should contain, but in practice it would probably work 100% well. The packages are always consumed by a bundler anyway.

One caveat could be with the `dist/types/index.d.ts` definitions. Normally the TS compiler needs to read only the type declarations of imported modules, that's all it needs to typecheck a compile a module. But now it would read the full source files. Which are much bigger.

> Node.js now has [support for loading TypeScript files](https://nodejs.org/api/typescript.html#type-stripping)?

Oh that's nice. It would be interesting to enable the `erasableSyntaxOnly` compiler option and see how compatible our TypeScript code is. Probably not much, we define plenty of `enum`s.

aduth replied:
> That would be breaking some formal/informal rules about what NPM packages should contain, but in practice it would probably work 100% well.

What I'm curious to understand is if those rules are (or should be) changing since Node.js now has native support for TypeScript.

My understanding is that it would previously been discouraged because it would require an additional build setup that would support interpreting the TypeScript source, but that's no longer the case.

> One caveat could be with the `dist/types/index.d.ts` definitions. Normally the TS compiler needs to read only the type declarations of imported modules, that's all it needs to typecheck a compile a module. But now it would read the full source files. Which are much bigger.

Yeah, type definitions are the main thing I wonder about in terms of the potential implications. In theory it seems that TypeScript and IDE integrations should be able to understand the types from source, but it's unclear to me if that's severely disadvantageous compared to publishing separate `.d.ts` files.

aduth replied:
> My understanding is that it would previously been discouraged because it would require an additional build setup that would support interpreting the TypeScript source, but that's no longer the case.

Although this applies to server-side code, not entirely clear if the same can be said about code that would end up being bundled for use in the browser.

jsnajdr reviewed (approved):
The `main: index.js` field in `package.json` should be either `module: index.js` (the file is an ES module) or `exports: index.js`. But that's a non blocker, `main` also works.

aduth committed:
polyfills: Correctly identify as ESM module

aduth committed:
polyfills: Add exports directive

aduth commented:
> The `main: index.js` field in `package.json` should be either `module: index.js` (the file is an ES module) or `exports: index.js`. But that's a non blocker, `main` also works.

Since this is an ES module as you noted, I think it makes sense to mark as such with `type` field, added in b5ec876. Adding `module` would work, but AFAIK that field was never really officially supported and hopefully we can move away from with `exports` instead.

Added `exports` in ca53952.
